File: src/game/conditions.js
/**
 * Condition helpers — pure game-state operations on figureConditions.
 * No Discord dependency.
 */
import { getDcEffects } from '../data-loader.js';
import { dcNameFromFigureKey } from './dc-helpers.js';
import { checkCapitalizePassiveRedraw } from './cc-passive-redraw.js';

import { getDcEffect } from './dc-helpers.js';
export const HARMFUL_CONDITIONS = ['Stun', 'Bleed', 'Weaken'];

/**
 * Remove a specific condition from a figure.
 * @param {object} game
 * @param {string} figureKey
 * @param {string} cond - condition name to remove (e.g. 'Stun', 'Bleed')
 */
export function filterCondition(game, figureKey, cond) {
  if (!game.figureConditions?.[figureKey]) return;
  // Disarm permanent Weakened: skip removal of Weaken if the figure has the Disarm lock
  if (cond === 'Weaken' && game.disarmPermanentWeakened?.[figureKey]) return;
  game.figureConditions[figureKey] = game.figureConditions[figureKey].filter((c) => c !== cond);
  if (game.figureConditions[figureKey].length === 0) delete game.figureConditions[figureKey];
}

/**
 * Check if a figure is immune to HARMFUL conditions (Stun, Bleed, Weaken).
 * Currently: Onar Koma, Snowtrooper Elite.
 *
 * Per destruct's 2026-05-05 audit, YWNDM-on-Fifth-Brother is NOT immunity:
 * the condition is APPLIED but INERT (token placed, effects suppressed —
 * matters for HK-style "while you have a HARMFUL condition" checks).
 * That semantic moved to areConditionEffectsSuppressed below.
 *
 * @param {object} game
 * @param {string} figureKey
 * @returns {boolean}
 */
export function isConditionImmune(game, figureKey) {
  const dcName = dcNameFromFigureKey(figureKey);
  const dcEff = getDcEffect(dcName);
  const sIds = dcEff?.specialAbilityIds || [];
  if (sIds.includes('immune_onar') || sIds.includes('immune_snowtrooper_elite')) return true;
  return false;
}

/**
 * Check whether a figure's HARMFUL condition effects are currently suppressed.
 * The conditions remain APPLIED on the figure (token in figureConditions);
 * only their downstream effects (Bleed end-of-action damage, Stun cannot-
 * declare-attack, Weaken −1 to attack/defense) should be skipped.
 *
 * Currently: Fifth Brother under YWNDM.
 *
 * Per destruct 2026-05-05: "condition is applied but inert. token IS placed
 * on the figure, effects suppressed. matters for HKs (HK abilities key off
 * having a condition — the condition still satisfies that check, just
 * doesn't damage you)."
 *
 * @param {object} game
 * @param {string} figureKey
 * @returns {boolean}
 */
/** YWNDM is per-player now — protect this figure only if ITS OWNER has YWNDM
 * active (a Fifth Brother whose own player played the card). alexanbv 2026-06-17. */
function ywndmActiveForFigure(game, figureKey) {
  const a = game?.youWillNotDenyMeActive;
  if (!a) return false;
  for (const pn of [1, 2]) {
    if (a[pn] && game.figurePositions?.[pn]?.[figureKey]) return true;
  }
  return false;
}

export function areConditionEffectsSuppressed(game, figureKey) {
  if (!ywndmActiveForFigure(game, figureKey)) return false;
  const dcName = dcNameFromFigureKey(figureKey);
  if (dcName?.toLowerCase().includes('fifth brother')) return true;
  return false;
}

/**
 * Check whether a figure cannot currently be defeated. Per destruct 2026-05-05:
 * "damage tokens capped at health. Pierce does NOT overflow."
 *
 * Sources today:
 *   - YWNDM (You Will Not Deny Me) on Fifth Brother
 *   - Second Chance CC attached to a specific DC msgId
 *   - Sustained by Rage on Krrsantan (when "no activation resolved this round")
 *
 * The destruct-correct semantic is: when this returns true, applying damage
 * that would defeat the figure should leave HP at 0 (or damage at health,
 * equivalently) and SUPPRESS the defeat trigger. Subsequent damage hits
 * with flag still active also stay at 0/no-defeat. A heal removes damage
 * normally and may bring the figure back above 0.
 *
 * This replaces the legacy "heal to 1 on defeat" approach which incorrectly
 * left figures vulnerable to a single-damage follow-up.
 *
 * @param {object} game
 * @param {string} figureKey
 * @param {string} [msgId] - optional, enables Second Chance check
 * @returns {boolean}
 */
export function isCannotBeDefeated(game, figureKey, msgId = null) {
  if (!game) return false;
  // YWNDM applies to Fifth Brother only — and only its owner's (per-player).
  if (ywndmActiveForFigure(game, figureKey)) {
    const dcName = dcNameFromFigureKey(figureKey);
    if (dcName?.toLowerCase().includes('fifth brother')) return true;
  }
  // Second Chance is keyed per msgId.
  if (msgId && game.secondChanceDcMsgId?.[msgId]) return true;
  // Sustained by Rage (Krrsantan, Maul): "If you have not resolved an
  // activation this round, you cannot be defeated." Per destruct
  // 2026-05-07. Hoisted from combat-bridge.js inline check so direct-
  // damage paths (Bleed end-of-action, Blast splash, mission rules,
  // self-damage abilities) also honor the protection — not just the
  // primary attack-resolution path.
  try {
    const dcName = dcNameFromFigureKey(figureKey);
    const eff = dcName ? _getSbrDcEff(dcName) : null;
    if ((eff?.specialAbilityIds || []).includes('sustained_by_rage')) {
      const pn = _findFigureOwner(game, figureKey);
      if (pn) {
        const dcList = pn === 1 ? game.p1DcList : game.p2DcList;
        const idx = Array.isArray(dcList) ? dcList.findIndex((dc) => dc?.dcName === dcName) : -1;
        const activatedIdxs = (pn === 1 ? game.p1ActivatedDcIndices : game.p2ActivatedDcIndices) || [];
        if (idx >= 0 && !activatedIdxs.includes(idx)) return true;
      }
    }
  } catch { /* defensive: fall through to false */ }
  return false;
}

function _getSbrDcEff(dcName) {
  return getDcEffects()?.[dcName] || null;
}

function _findFigureOwner(game, figureKey) {
  for (const pn of [1, 2]) {
    if (game.figurePositions?.[pn]?.[figureKey] != null) return pn;
  }
  return null;
}

/**
 * Apply a condition to a figure (with dedup). Initialises figureConditions if needed.
 * @param {object} game
 * @param {string} figureKey
 * @param {string} cond - condition name (e.g. 'Focus', 'Stun', 'Hide')
 * @returns {boolean} true if the condition was newly applied, false if already present
 */
/**
 * True if the figure is Vanished (cannot receive new conditions / suffer Damage
 * until its next activation). Existing conditions are NOT affected. Checks both
 * players' Vanish records (alexanbv 2026-06-19).
 */
export function isFigureVanished(game, figureKey) {
  for (const pn of [1, 2]) {
    const v = game?.vanishImmunityUntilNextActivation?.[pn];
    if (!v) continue;
    if (v.figureKey ? v.figureKey === figureKey : (v.dcName && String(figureKey).startsWith(`${v.dcName}-`))) return true;
  }
  return false;
}

export function applyCondition(game, figureKey, cond) {
  // CRR-INCP-002: an incapacitated figure cannot have conditions applied.
  // Skirmish substrate: The Child while game.childIncapacitated is true.
  if (game?.childIncapacitated && dcNameFromFigureKey(figureKey) === 'The Child') return false;
  // Vanish: the figure cannot RECEIVE new conditions (harmful or beneficial)
  // until its next activation; conditions already present remain untouched.
  if (isFigureVanished(game, figureKey)) return false;
  game.figureConditions = game.figureConditions || {};
  game.figureConditions[figureKey] = game.figureConditions[figureKey] || [];
  if (game.figureConditions[figureKey].includes(cond)) return false;
  // Capitalize passive trigger (per CRR 2026-05-09): if this is a
  // HARMFUL condition AND the figure already has at least one OTHER
  // harmful condition, fire the Capitalize redraw for the figure's
  // OPPONENT (the player who would own Capitalize). Only fires once
  // per stacking event; the actual redraw lives in cc-passive-redraw.
  const _capIsHarmful = HARMFUL_CONDITIONS.includes(cond);
  const _capPriorHarmful = _capIsHarmful && game.figureConditions[figureKey].some((c) => HARMFUL_CONDITIONS.includes(c));
  game.figureConditions[figureKey].push(cond);
  if (_capPriorHarmful) {
    const _capOwnerPN = game.figurePositions?.[1]?.[figureKey] != null ? 1
      : game.figurePositions?.[2]?.[figureKey] != null ? 2 : null;
    if (_capOwnerPN) checkCapitalizePassiveRedraw(game, figureKey, _capOwnerPN);
  }
  return true;
}

/**
 * Ensure a condition is set on a figure, replacing any existing instance.
 * Useful when a condition must be present exactly once regardless of prior state.
 * @param {object} game
 * @param {string} figureKey
 * @param {string} cond
 */
/**
 * Apply a condition and, if it stuck, append a bonus die to attackInfo.
 * Encodes the rule: "become [Condition] → gain +1 [die color]".
 * @param {object} game
 * @param {string} figureKey
 * @param {string} condition - e.g. 'Focus'
 * @param {object} attackInfo - current attack info (has .dice array)
 * @param {string} dieColor - e.g. 'green'
 * @returns {{ attackInfo: object, applied: boolean }}
 */
export function applyConditionWithDie(game, figureKey, condition, attackInfo, dieColor) {
  if (applyCondition(game, figureKey, condition)) {
    return {
      attackInfo: { ...attackInfo, dice: [...(attackInfo.dice || []), dieColor] },
      applied: true,
    };
  }
  return { attackInfo, applied: false };
}

export function resetCondition(game, figureKey, cond) {
  game.figureConditions = game.figureConditions || {};
  game.figureConditions[figureKey] = [...(game.figureConditions[figureKey] || []).filter(c => c !== cond), cond];
}

File: python/engine/mechanics/conditions.py
"""D2.24 + D2.25 — condition helpers.

Byte-identical port of `src/game/conditions.js`. Pure game-state operations on
`game.figureConditions` — no Discord dependency, no handler UI.

Conditions model the dedup + Weaken-lock rules from the JS engine. The IA
rulebook lists nine named conditions; only five are actually implemented in the
JS engine today (**Stun**, **Bleed**, **Focus**, **Hide**, **Weaken**). The
remaining four (Immobilize, Exposed, Disoriented, Hindered) are reserved
names — none of the JS code ever applies or reads them. The Python port mirrors
that reality: it's a string bag keyed on figureKey, no per-condition typing.
Apply any of the five real labels via `apply_condition(game, fk, 'Stun')`.

State shape:
    game['figureConditions']        dict[str, list[str]]  # figureKey → [cond, ...]
    game['disarmPermanentWeakened'] dict[str, bool]       # lock: Weaken can't be removed
    game['youWillNotDenyMeActive']  bool                  # Fifth Brother passive flag

D2.25 is the same module: `is_condition_immune` reads DC specialAbilityIds to
gate Onar Koma / Snowtrooper Elite / Fifth Brother (under YWNDM) from harmful
conditions.
"""
from __future__ import annotations

from typing import Any, Dict, Optional, Tuple

from python.engine.data.dc_effects_loader import get_dc_effect
from python.engine.mechanics.dc_helpers import dc_name_from_figure_key

# ── Constants ───────────────────────────────────────────────────────────────

HARMFUL_CONDITIONS: Tuple[str, ...] = ('Stun', 'Bleed', 'Weaken')
"""Harmful conditions gated by `is_condition_immune`.

Mirrors `src/game/conditions.js:HARMFUL_CONDITIONS`. Focus and Hide are
beneficial and never gated."""


# ── D2.24 — applyCondition / applyConditionWithDie / filterCondition / reset

def apply_condition(game: Dict[str, Any], figure_key: str, cond: str) -> bool:
    """Apply a condition to a figure with dedup.

    Mirrors `conditions.js:applyCondition`. Initialises `figureConditions` and
    the per-figure list if missing. Returns True iff the condition was newly
    applied; False if already present on the figure.
    """
    game.setdefault('figureConditions', {})
    game['figureConditions'].setdefault(figure_key, [])
    if cond in game['figureConditions'][figure_key]:
        return False
    game['figureConditions'][figure_key].append(cond)
    return True


def apply_condition_with_die(game: Dict[str, Any],
                             figure_key: str,
                             condition: str,
                             attack_info: Dict[str, Any],
                             die_color: str) -> Dict[str, Any]:
    """Apply a condition and, if newly applied, append a bonus die to attack_info.

    Encodes the rule "become [Condition] → gain +1 [die color]". Mirrors
    `conditions.js:applyConditionWithDie`. Returns `{attackInfo, applied}`.
    The returned `attackInfo` is a shallow copy with an extended `dice` list
    — the input dict is not mutated (matches JS spread semantics).
    """
    if apply_condition(game, figure_key, condition):
        new_dice = list(attack_info.get('dice') or []) + [die_color]
        new_info = {**attack_info, 'dice': new_dice}
        return {'attackInfo': new_info, 'applied': True}
    return {'attackInfo': attack_info, 'applied': False}


def filter_condition(game: Dict[str, Any], figure_key: str, cond: str) -> None:
    """Remove a specific condition from a figure. No-op if not present.

    Respects the disarmPermanentWeakened lock: if `cond == 'Weaken'` and the
    figure has the Disarm lock, the call is a no-op. If the per-figure list
    becomes empty after removal, the key is deleted from `figureConditions` —
    matches the JS `delete` branch so round-trip equality with JS-produced
    state is preserved.
    """
    fc = game.get('figureConditions')
    if not fc or figure_key not in fc:
        return
    # Disarm permanent Weakened: skip removal of Weaken if the figure has the lock.
    if cond == 'Weaken':
        lock = game.get('disarmPermanentWeakened') or {}
        if lock.get(figure_key):
            return
    fc[figure_key] = [c for c in fc[figure_key] if c != cond]
    if not fc[figure_key]:
        del fc[figure_key]


def reset_condition(game: Dict[str, Any], figure_key: str, cond: str) -> None:
    """Ensure a condition is present exactly once on a figure.

    Removes any existing copies and appends once. Useful when a condition must
    be set regardless of prior state. Mirrors `conditions.js:resetCondition`.
    """
    game.setdefault('figureConditions', {})
    existing = game['figureConditions'].get(figure_key) or []
    game['figureConditions'][figure_key] = [c for c in existing if c != cond] + [cond]


# ── D2.25 — is_condition_immune ─────────────────────────────────────────────

def is_condition_immune(game: Dict[str, Any], figure_key: str) -> bool:
    """Check if a figure is immune to HARMFUL conditions (Stun/Bleed/Weaken).

    Mirrors `conditions.js:isConditionImmune`. Dynamic gates:
      - `immune_onar` in DC specialAbilityIds (Onar Koma).
      - `immune_snowtrooper_elite` in DC specialAbilityIds.
      - `game.youWillNotDenyMeActive` AND DC name contains "fifth brother"
        (Fifth Brother passive while YWNDM active).

    Returns False for any figure whose DC has none of the above.
    """
    dc_name = dc_name_from_figure_key(figure_key)
    dc_eff = get_dc_effect(dc_name) or {}
    s_ids = dc_eff.get('specialAbilityIds') or []
    if 'immune_onar' in s_ids or 'immune_snowtrooper_elite' in s_ids:
        return True
    if game.get('youWillNotDenyMeActive') and dc_name and 'fifth brother' in dc_name.lower():
        return True
    return False
